main/src/ca/uwaterloo/flix/language/phase/optimizer/RecursionRewriter.scala

```diff
@@ -0,0 +1,454 @@
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.ApplyClo(exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos)
+      val t2 = visitExp(exp2, tailPos = TailPosition.NonTail)
+      TailPosition.combine(t1, t2)
+
+    case Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc) =>
+      // Check for recursion
+      if (sym == sym0) {
+        if (tailPos == TailPosition.NonTail) {
+          return tailPos
+        }
+        ctx.selfTailCalls.addOne(Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc))
+      }
+      // Check alive parameters
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.ApplyLocalDef(_, exps, _, _, _) =>
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.Let(_, exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      TailPosition.combine(t1, t2)
+
+    case Expr.LocalDef(_, _, exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      TailPosition.combine(t1, t2)
+
+    case Expr.Scope(_, _, exp, _, _, _) =>
+      visitExp(exp, tailPos)
+
+    case Expr.IfThenElse(exp1, exp2, exp3, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      val t3 = visitExp(exp3, tailPos)
+      TailPosition.combine(TailPosition.combine(t1, t2), t3)
+
+    case Expr.Stm(exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos)
+      TailPosition.combine(t1, t2)
+
+    case Expr.Discard(exp, _, _) =>
+      visitExp(exp, tailPos)
+
+    case Expr.Match(exp1, rules, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = rules.map {
+          case MonoAst.MatchRule(_, guard, exp2) =>
+            val gt1 = guard.map(visitExp(_, tailPos = TailPosition.NonTail))
+            val et2 = visitExp(exp2, tailPos)
+            TailPosition.combineOpt(gt1, et2)
+        }.reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+      TailPosition.combine(t1, t2)
+
+    case Expr.VectorLit(exps, _, _, _) =>
+      exps.map(visitExp(_, tailPos = TailPosition.NonTail))
+        .reduceOption(TailPosition.combine)
+        .getOrElse(tailPos)
+
+    case Expr.VectorLoad(exp1, exp2, _, _, _) =>
+      val t1 = visitExp(exp1, tailPos = TailPosition.NonTail)
+      val t2 = visitExp(exp2, tailPos = TailPosition.NonTail)
+      TailPosition.combine(t1, t2)
+
+    case Expr.VectorLength(exp, _) =>
+      visitExp(exp, tailPos = TailPosition.NonTail)
+
+    case Expr.Ascribe(exp, _, _, _) => // TODO: Is this erased???
```

Comment:
What AST is this?

Reply:
MonoAst

Reply:
Neither Ascribe nor Cast is unused... so I am not sure what you mean

Reply:
This is confirmed by throwing ICE and running tests

Reply:
> Neither Ascribe nor Cast is unused... so I am not sure what you mean

Used or unused?

Reply:
They occur in real programs

Reply:
What I mean by erasure is that it is erased later so at runtime the expression would be in tailposition. Ascribe and Cast only carry type information the compiler needs but should not really have an impact at runtime.

Reply:
No, don't mess with this. Don't make such assumptions.

Reply:
Does not matter anymore